core/src/main/java/org/apache/iceberg/TableMetadataParser.java

```diff
@@ -252,22 +248,7 @@ private static void toJson(TableMetadata metadata, JsonGenerator generator) thro
   private static void toJson(Map<String, SnapshotRef> refs, JsonGenerator generator) throws IOException {
     generator.writeObjectFieldStart(REFS);
     for (Map.Entry<String, SnapshotRef> refEntry : refs.entrySet()) {
-      generator.writeObjectFieldStart(refEntry.getKey());
-
-      SnapshotRef ref = refEntry.getValue();
-      generator.writeStringField(TYPE, ref.type().toString().toLowerCase(Locale.ROOT));
-      generator.writeNumberField(SNAPSHOT_ID, ref.snapshotId());
-      if (ref.maxRefAgeMs() != null) {
-        generator.writeNumberField(MAX_REF_AGE_MS, ref.maxRefAgeMs());
-      }
-      if (ref.maxSnapshotAgeMs() != null) {
-        generator.writeNumberField(MAX_SNAPSHOT_AGE_MS, ref.maxSnapshotAgeMs());
-      }
-      if (ref.minSnapshotsToKeep() != null) {
-        generator.writeNumberField(MIN_SNAPSHOTS_TO_KEEP, ref.minSnapshotsToKeep());
-      }
-
-      generator.writeEndObject();
+      SnapshotRefParser.toJson(refEntry.getKey(), refEntry.getValue(), generator);
```

Comment:
I think that the `writeObjectFieldStart` and `writeEndObject` should be here in this function. Moving it to `toJson` in the other class creates an implicit requirement on the state of the `JsonGenerator`. It _must_ be writing an object. We avoid depending on the state of objects passed in where possible, **especially** for methods that are public.

Instead, the state of the generator should always be that it is ready to write any JSON value.

Reply:
Updated. We can write the name field and leverage the existing toJson in SnapshotRefParser. I think this approach works because the ref can be serialized independently without any reliance on state. In TableMetadataParser we can serialize the name + ref entity together